python/ray/serve/storage/kv_store.py

```diff
@@ -0,0 +1,251 @@
+                       "(key TEXT UNIQUE, value BLOB)")
+        self._conn.commit()
+
+    def get_storage_key(self, key):
+        return "{ns}-{key}".format(ns=self._namespace, key=key)
+
+    def put(self, key, val):
+        """Put the key-value pair into the store.
+
+        Args:
+            key (str)
+            val (bytes)
+        """
+        if not isinstance(key, str):
+            raise TypeError(f"key must be a string, got: {type(key)}.")
+        if not isinstance(val, bytes):
+            raise TypeError(f"val must be bytes, got: {type(val)}.")
+
+        cursor = self._conn.cursor()
+        cursor.execute(
+            f"INSERT OR REPLACE INTO {self._namespace} "
+            "(key, value) VALUES (?,?)", (self.get_storage_key(key), val))
+        self._conn.commit()
+
+    def get(self, key):
+        """Get the value associated with the given key from the store.
+
+        Args:
+            key (str)
+
+        Returns:
+            The bytes value. If the key wasn't found, returns None.
+        """
+        if not isinstance(key, str):
+            raise TypeError(f"key must be a string, got: {type(key)}.")
+
+        cursor = self._conn.cursor()
+        result = list(
+            cursor.execute(
+                f"SELECT value FROM {self._namespace} WHERE key = (?)",
+                (self.get_storage_key(key), )))
+        if len(result) == 0:
+            return None
+        else:
+            # Due to UNIQUE constraint, there can only be one value.
+            value, *_ = result[0]
+            return value
+
+    def delete(self, key):
+        """Delete the value associated with the given key from the store.
+
+        Args:
+            key (str)
+        """
+
+        if not isinstance(key, str):
+            raise TypeError("key must be a string, got: {}.".format(type(key)))
+
+        cursor = self._conn.cursor()
+        cursor.execute(f"DELETE FROM {self._namespace} "
+                       "WHERE key = (?)", (self.get_storage_key(key), ))
+        self._conn.commit()
+
+
+class RayS3KVStore(KVStoreBase):
+    """Persistent version of KVStoreBase for cluster fault
+    tolerance. Writes to S3 bucket with provided path and credentials.
+
+    Supports only string type for key and bytes type for value,
+    caller must handle serialization.
+    """
+
+    def __init__(
+            self,
+            namepsace: str,
+            bucket="",
+            s3_path="",
+            region_name="us-west-2",
+            aws_access_key_id=None,
+            aws_secret_access_key=None,
+            aws_session_token=None,
+    ):
+        self._namespace = namepsace
+        self._bucket = bucket
+        self._s3_path = s3_path
+
+        self._s3 = boto3.client(
+            "s3",
+            region_name=region_name,
+            aws_access_key_id=aws_access_key_id,
+            aws_secret_access_key=aws_secret_access_key,
+            aws_session_token=aws_session_token)
```

Comment:
will these be read from the filesystem  / env variables if not provided?

Reply:
i left it to user code (in our case, unit test input) to decide but it works either way